File: algorithms/03_cpp/math/extended_gcd.cpp
/**
 * Extended Euclidean Algorithm in C++ (CLRS 3rd Ed. Chapter 31.2)
 */

#include <iostream>
#include <tuple>
#include <cassert>

std::tuple<long long, long long, long long> extendedGCD(long long a, long long b) {
    if (b == 0) return {a, 1, 0};
    auto [gcd, x1, y1] = extendedGCD(b, a % b);
    long long x = y1;
    long long y = x1 - (a / b) * y1;
    return {gcd, x, y};
}

int main() {
    auto [g, x, y] = extendedGCD(240, 46);
    assert(g == 2);
    assert(240 * x + 46 * y == 2);
    std::cout << "C++ Extended GCD verified.\n";
    return 0;
}

File: algorithms/11_golang/math/extended_gcd.go
// Package math implements Extended Euclidean Algorithm (CLRS 3rd Ed. Chapter 31.2).
package math

func ExtendedGCD(a, b int64) (gcd, x, y int64) {
	if b == 0 {
		return a, 1, 0
	}
	gcd, x1, y1 := ExtendedGCD(b, a%b)
	x = y1
	y = x1 - (a/b)*y1
	return gcd, x, y
}
